File: typescript/libs/workspace/adapters/system/src/lib/session/originstate/DatabaseOriginState.ts
import { Operations } from '@allors/workspace/domain/system';
import { MethodType, RoleType } from '@allors/workspace/meta/system';
import { DatabaseRecord } from '../../database/DatabaseRecord';
import { IRecord } from '../../IRecord';
import { Strategy } from '../Strategy';
import { RecordBasedOriginState } from './RecordBasedOriginState';

export const UnknownVersion = 0;
export const InitialVersion = 1;

export abstract class DatabaseOriginState extends RecordBasedOriginState {
  databaseRecord: DatabaseRecord;

  protected constructor(strategy: Strategy, record: DatabaseRecord) {
    super(strategy);
    this.databaseRecord = record;
    this.previousRecord = this.databaseRecord;
  }

  public get version(): number {
    return this.databaseRecord?.version ?? UnknownVersion;
  }

  private get isVersionUnknown(): boolean {
    return this.version == UnknownVersion;
  }

  get roleTypes(): Set<RoleType> {
    return this.class.databaseOriginRoleTypes;
  }

  protected get existRecord(): boolean {
    return this.record != null;
  }

  get record(): IRecord {
    return this.databaseRecord;
  }

  public canRead(roleType: RoleType): boolean {
    if (!this.existRecord) {
      return true;
    }

    if (this.isVersionUnknown) {
      return false;
    }

    const permission = this.session.workspace.database.getPermission(this.class, roleType, Operations.Read);
    return this.databaseRecord.isPermitted(permission);
  }

  public canWrite(roleType: RoleType): boolean {
    if (!this.existRecord) {
      return true;
    }

    if (this.isVersionUnknown) {
      return false;
    }

    const permission = this.session.workspace.database.getPermission(this.class, roleType, Operations.Write);
    return this.databaseRecord.isPermitted(permission);
  }

  public canExecute(methodType: MethodType): boolean {
    if (!this.existRecord) {
      return true;
    }

    if (this.isVersionUnknown) {
      return false;
    }

    const permission = this.session.workspace.database.getPermission(this.class, methodType, Operations.Execute);
    return this.databaseRecord.isPermitted(permission);
  }

  public pushResponse(newDatabaseRecord: DatabaseRecord) {
    this.databaseRecord = newDatabaseRecord;
    this.changedRoleByRelationType = null;
  }

  public onPulled() {
    this.databaseRecord = this.session.workspace.database.getRecord(this.id);
  }

  onChange() {
    this.session.changeSetTracker.onDatabaseChanged(this);
    this.session.pushToDatabaseTracker.onChanged(this);
  }
}

File: dotnet/System/workspace/csharp/adapters/allors.workspace.adapters/session/originstate/DatabaseOriginState.cs
// <copyright file="DatabaseOriginState.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System.Collections.Generic;
    using Meta;

    public abstract class DatabaseOriginState : RecordBasedOriginState
    {
        protected DatabaseOriginState(DatabaseRecord record)
        {
            this.DatabaseRecord = record;
            this.PreviousRecord = this.DatabaseRecord;
            this.IsPushed = false;
        }

        public long Version => this.DatabaseRecord?.Version ?? Allors.Version.WorkspaceInitial;

        internal bool IsVersionInitial => this.Version == Allors.Version.WorkspaceInitial.Value;

        protected override IEnumerable<IRoleType> RoleTypes => this.Class.DatabaseOriginRoleTypes;

        protected bool ExistRecord => this.Record != null;

        protected override IRecord Record => this.DatabaseRecord;

        protected internal DatabaseRecord DatabaseRecord { get; private set; }

        public bool IsPushed { get; private set; }

        public bool CanRead(IRoleType roleType)
        {
            if (!this.ExistRecord)
            {
                return true;
            }

            if (this.IsVersionInitial)
            {
                // TODO: 
                return true;
            }

            var permission = this.Session.Workspace.DatabaseConnection.GetPermission(this.Class, roleType, Operations.Read);
            return this.DatabaseRecord.IsPermitted(permission);
        }

        public bool CanWrite(IRoleType roleType)
        {
            if (this.IsVersionInitial)
            {
                return !this.IsPushed;
            }

            if (this.IsPushed)
            {
                return false;
            }

            if (!this.ExistRecord)
            {
                return true;
            }

            var permission = this.Session.Workspace.DatabaseConnection.GetPermission(this.Class, roleType, Operations.Write);
            return this.DatabaseRecord.IsPermitted(permission);
        }

        public bool CanExecute(IMethodType methodType)
        {
            if (!this.ExistRecord)
            {
                return true;
            }

            if (this.IsVersionInitial)
            {
                // TODO: 
                return true;
            }

            var permission = this.Session.Workspace.DatabaseConnection.GetPermission(this.Class, methodType, Operations.Execute);
            return this.DatabaseRecord.IsPermitted(permission);
        }

        public void OnPushed() => this.IsPushed = true;

        public void OnPulled(IPullResultInternals result)
        {
            if (this.ChangedRoleByRelationType?.Count > 0)
            {
                result.AddMergeError(this.Strategy.Object);
                return;
            }

            this.IsPushed = false;
            this.DatabaseRecord = this.Session.Workspace.DatabaseConnection.GetRecord(this.Id);
        }

        protected override void OnChange()
        {
            this.Session.ChangeSetTracker.OnDatabaseChanged(this);
            this.Session.PushToDatabaseTracker.OnChanged(this);
        }
    }
}
